src/utils/structs/checkpoint.cairo

```diff
@@ -0,0 +1,242 @@
+#[derive(Copy, Drop, Serde)]
+struct Checkpoint {
+    key: u64,
+    value: u256
+}
+
+#[generate_trait]
+impl TraceImpl of TraceTrait {
+    /// Pushes a (`key`, `value`) pair into a Trace so that it is stored as the checkpoint
+    /// and returns previous value and new value.
+    fn push(ref self: Trace, key: u64, value: u256) -> (u256, u256) {
+        self.checkpoints._insert(key, value)
+    }
+
+    /// Returns the value in the last (most recent) checkpoint with key lower or equal
+    /// than the search key, or zero if there is none.
+    fn upper_lookup(self: @Trace, key: u64) -> u256 {
+        let checkpoints = self.checkpoints;
+        let len = checkpoints.len();
+        let pos = checkpoints._upper_binary_lookup(key, 0, len);
+
+        if pos == 0 {
+            0
+        } else {
+            checkpoints.read_at(pos - 1).value
+        }
+    }
+
+    /// Returns the value in the last (most recent) checkpoint with key lower or equal
+    /// than the search key, or zero if there is none.
+    ///
+    /// NOTE: This is a variant of {upper_lookup} that is optimised to
+    /// find "recent" checkpoint (checkpoints with high keys).
+    fn upper_lookup_recent(self: @Trace, key: u64) -> u256 {
+        let checkpoints = self.checkpoints;
+        let len = checkpoints.len();
+
+        let mut low = 0;
+        let mut high = len;
+
+        if (len > 5) {
+            let mid = len - u32_sqrt(len).into();
+            if (key < checkpoints.read_at(mid).key) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        let pos = checkpoints._upper_binary_lookup(key, low, high);
+
+        if pos == 0 {
+            0
+        } else {
+            checkpoints.read_at(pos - 1).value
+        }
+    }
+
+    /// Returns the value in the most recent checkpoint, or zero if there are no checkpoints.
+    fn latest(self: @Trace) -> u256 {
+        let checkpoints = self.checkpoints;
+        let pos = checkpoints.len();
+
+        if pos == 0 {
+            0
+        } else {
+            checkpoints.read_at(pos - 1).value
+        }
+    }
+
+    /// Returns whether there is a checkpoint in the structure (i.e. it is not empty),
+    /// and if so the key and value in the most recent checkpoint.
+    fn latest_checkpoint(self: @Trace) -> (bool, u64, u256) {
+        let checkpoints = self.checkpoints;
+        let pos = checkpoints.len();
+
+        if (pos == 0) {
+            (false, 0, 0)
+        } else {
+            let checkpoint: Checkpoint = checkpoints.read_at(pos - 1);
+            (true, checkpoint.key, checkpoint.value)
+        }
+    }
+
+    /// Returns the number of checkpoints.
+    fn length(self: @Trace) -> u32 {
+        self.checkpoints.len()
+    }
+
+    /// Returns the checkpoint at given position.
+    fn at(self: @Trace, pos: u32) -> Checkpoint {
+        assert(pos < self.length(), 'Array overflow');
+        self.checkpoints.read_at(pos)
+    }
```

Comment:
We have time to add the unit tests before merging this PR. I will work on it.

Reply:
I was working on this already, but I realized that these unit tests will most likely change when the native StorageArray (that Ariel mentioned is coming) arrives, at that point we will probably remove our own StorageArray implementation, and refactor Trace and Checkpoint from this. I lean toward creating the issue and waiting for the new storage array to come before adding the unit tests, since the module is already released, audited, and used in production. Wdyt?